File: dsa/concepts/13_graphs/java/Dfs.java

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class Dfs {

	public static void main(String[] args) {
		Dfs obj = new Dfs();
		List<List<Integer>> adj = new ArrayList<>();
		for (int i = 0; i < 4; i++)
			adj.add(new ArrayList<>());

		adj.get(0).add(1);
		adj.get(0).add(2);
		adj.get(1).add(2);
		adj.get(2).add(0);
		adj.get(2).add(3);
		adj.get(3).add(3);

		System.out.println(obj.dfs2(adj, 0));
		System.out.println(obj.dfs(adj, 0));
	}

	// Space: O(V)
	// Time: O(V + E)
	// Recursive approach
	private List<Integer> dfs(List<List<Integer>> adj, int src) {
		int n = adj.size();
		List<Integer> ans = new ArrayList<>();
		int[] vis = new int[n];
		dfs(adj, vis, 0, ans);

		return ans;
	}

	private void dfs(List<List<Integer>> adj, int[] vis, int cur, List<Integer> ans) {
		vis[cur] = 1;
		ans.add(cur);
		for (int nei : adj.get(cur)) {
			if (vis[nei] != 1) {
				dfs(adj, vis, nei, ans);
			}
		}
	}

	// Iterative approach
	public List<Integer> dfs2(List<List<Integer>> adj, int src) {
		int n = adj.size();
		List<Integer> ans = new ArrayList<>();

		int[] vis = new int[n];
		Stack<Integer> s = new Stack<>();

		s.add(src);
		vis[src] = 1;
		while (!s.isEmpty()) {
			int node = s.pop();
			ans.add(node);
			for (int nei : adj.get(node)) {
				if (vis[nei] != 1) {
					vis[nei] = 1;
					s.add(nei);
				}
			}
		}

		return ans;
	}

}

File: dsa/patterns/13_graphs/python/dfs.py
"""
Depth First Search (DFS)

- Graph is represented using adjacency list
- Supports:
  1) Recursive DFS
  2) Iterative DFS using stack

Time Complexity: O(V + E)
Space Complexity: O(V)
"""


def dfs_recursive(adj, src):
    """
    Recursive DFS
    """
    n = len(adj)
    visited = [0] * n
    result = []

    _dfs(adj, src, visited, result)
    return result


def _dfs(adj, node, visited, result):
    visited[node] = 1
    result.append(node)

    for nei in adj[node]:
        if visited[nei] == 0:
            _dfs(adj, nei, visited, result)


def dfs_iterative(adj, src):
    """
    Iterative DFS using stack
    """
    n = len(adj)
    visited = [0] * n
    result = []

    stack = []
    stack.append(src)
    visited[src] = 1

    while stack:
        node = stack.pop()
        result.append(node)

        for nei in adj[node]:
            if visited[nei] == 0:
                visited[nei] = 1
                stack.append(nei)

    return result


# -------------------- TEST --------------------
if __name__ == "__main__":
    adj = [[] for _ in range(4)]

    adj[0].append(1)
    adj[0].append(2)
    adj[1].append(2)
    adj[2].append(0)
    adj[2].append(3)
    adj[3].append(3)

    print("Iterative DFS:", dfs_iterative(adj, 0))
    print("Recursive DFS:", dfs_recursive(adj, 0))
